### DISCUSSION TITLE
Question: how far can I get building a (packaged?) C++ and WinUI 3 app, using the command-line only?

### DISCUSSION THREAD

**Discussion by galexite** [microsoft/WindowsAppSDK#1176] [Q&A]
Hi there! :wave:

I really like the new Windows App SDK and the associated WinUI 3, and C++/WinRT projects, and I've been toying with a few apps using the libraries. However, one thing I'm not so clear about, is how I can build a Windows App SDK project without using Visual Studio.

I'm not opposed to using Visual Studio _per se_, but I'd like to bring over the tools and workflow I'm already familiar with on Mac and Linux, which is also available to Win32 apps -- CMake, Visual Studio Code, CLion and the like. If I were building an app, I'd also like to be able to maintain a single, consistent build system for apps and libraries which share cross-platform business logic.

I'd like to create a small CMake module or starter project which abstracts away most of the boilerplate required to build a WinUI 3 app using just CMake. I realise that most of the work is done when generating a Visual Studio project, but is it possible to use just only command-line tools? This way, a CMake project can be built upon custom targets, therefore allowing the user to choose another build system, such as a Makefile, or Ninja.

Many thanks,
George

**Comment by galexite** [comment]
It looks like you should be able to straight away build a WinUI 3 app unpackaged, using [MSIX Dynamic Dependencies](https://github.com/microsoft/WindowsAppSDK/blob/main/specs/dynamicdependencies/DynamicDependencies.md) as [microsoft/windows-app-rs](https://github.com/microsoft/windows-app-rs) does. I suppose the next issue is compiling XAML files.

**Reply by riverar** [reply]
Oh hey, I created `windows-app-rs` crate. Dynamic Dependencies will be fully supported in 1.0 (and should land in the next preview or so). It's not supported at all in 0.8-preview so I don't exactly recommend doing what the `windows-app-rs` crate does just yet 😂

**Reply by galexite** [reply]
Hi @riverar! Thanks for the `windows-app-rs` crate! I understand, I'm not going to be writing any production apps any time soon, I'm just interested in playing around with getting CMake going. I realise unpackaged apps aren't supported until the 1.0 release later this year, but the samples from `windows-samples-rs` work quite well already.

**Comment by wjk** [comment]
@galexite You can very much build a WinUI app via the command line only, but you’ll need to write a good chunk of support code.

* makepri.exe, midl.exe, and mdmerge.exe will need to be run under Wine. While Wine has a similar program called widl, don’t try using it unless you’re willing to invest a significant amount of time adding the required features.
* CMake can download NuGet packages from the Internet at configure time or build time, and the direct download links of packages on the public NuGet feed are published on that package’s webpage. CMake can also extract zip files at configure time or build time, and a NuGet package is just a zipfile. No special tooling is required here.
* The standalone App SDK installer is all you need to ship for dependencies (just run it elevated and it will do all the hard work), unless you are using MSIX. In that case, you must bundle the MSIX packages containing the dependencies and ensure that they are installed first. I have no clue how to do this, even when using Visual Studio.
* Regarding the XBF compiler. You do not need to use XBF at all if you can generate equivalent imperative C++ code like what Windows Forms does with its designer. While an automated generator does not yet exist, all of the hard work of creating one has already been done. As a bonus, your UI code would become nigh-impossible to reverse engineer because it’s just C++; no blob is required.
* There is an API that can parse a string containing human-editable XAML and create the objects specified in that XAML at runtime: XamlReader.Load.
* The code-behind files are only an artifact of the Visual Studio design model. You are free to load, store, and use your XAML however you want.
* And when you’re done, use makemsix to create (but not currently sign) an MSIX package; Wine is not required here.

Enjoy! (updated 2021-01-07)

**Comment by sotanakamura** [comment]
See [Build C++ WinUI 3 Apps with Command Line Tools](https://github.com/sotanakamura/winui3-command-line) or [MinGW WinUI 3 Demo](https://github.com/driver1998/WinUI-MinGW)
